Serialisation and Message introspection

ch3pjw opened the pull request:
This PR adds some introspection to Message objects so one can see what args they contain (useful for debugging) and exposes the message serialisation/deserialisation (useful for applications that don't actually want to send messages via `liblo`'s server, but do want to use its message processing).

Caveat: I'm quite new to Cython, so please check this carefully! Especially around the binary blob handing - mostly my tests pass, but sometimes they don't and I get strange trailing data in the blob.

ch3pjw committed:
Add lo_message_serialise to liblo C declarations

ch3pjw committed:
Expose lo_message_serialise() as liblo.Message.serialise()

foolswood committed:
Free intermediary C string

foolswood committed:
Add message deserialisation

ch3pjw committed:
Make Message object more introspectable...

by exposing various lo_message methods conveniently. Also, factor out the arg
processing for re-use.

ch3pjw committed:
Add tests for new Message methods

ch3pjw commented:
Is this kind of thing useful? Any chance of having a quick once over?

ch3pjw commented:
Any chance of looking over this?

dsacre commented:
I've finally had some time to look at the code, seems pretty good to me.

I managed to reproduce the test failures you mentioned, and I think they are due to a bug that has existed ever since I added Python 3 support to pyliblo. Please try if commit c53d8e2f61e03398d7372a1ecff320f97c2a0acb fixes the remaining issues for you.

`MessageTestCase.testArgs()` still fails in Python 2 because blobs are returned as lists, but that's easily fixed.

Could you add docstrings for the newly added functions and attributes, and include them in `doc/index.rst`?

ch3pjw commented:
Fab, thanks! Yes, I'll have a look at doing the docs, and I'll try out your fix.

If you're happy for me too, I was thinking of adding the same kind of introspection for bundles at some point, though it looks a little more involved.

ch3pjw committed:
Document Message.{deserialise,serialise}

ch3pjw committed:
Document Message.{path,args,types}...

and convert them to using the more canonical cython property syntax

ch3pjw committed:
Fix MessageTestCase WRT Python 2 byte handling

ch3pjw commented:
I've updated the docs, hope they're to your liking, and I've fixed that test for Python 2. As far as one can tell with intermittent test failures your fix appears good.

ch3pjw commented:
@dsacre is this good to go now?
